src/hotspot/share/logging/logAsyncFlusher.cpp

```diff
@@ -0,0 +1,233 @@
+    enqueue_impl(m);
+  }
+}
+
+LogAsyncFlusher::LogAsyncFlusher()
+  : _state(ThreadState::Running),
+    _lock(Mutex::tty, "async-log-monitor", true /* allow_vm_block */, Mutex::_safepoint_check_never),
+    _stats(17 /*table_size*/) {
+  if (os::create_thread(this, os::asynclog_thread)) {
+    os::start_thread(this);
+  }
+
+  log_info(logging)("The maximum entries of AsyncLogBuffer: " SIZE_FORMAT ", estimated memory use: " SIZE_FORMAT " bytes",
+                    _buffer_max_size, AsyncLogBufferSize);
+}
+
+bool AsyncLogMapIterator::do_entry(LogFileOutput* output, uintx* counter) {
+  using dummy = LogTagSetMapping<LogTag::__NO_TAG>;
+  LogDecorations decorations(LogLevel::Warning, dummy::tagset(), output->decorators());
+  const int sz = 128;
+  char out_of_band[sz];
+
+  if (*counter > 0) {
+    jio_snprintf(out_of_band, sz, UINTX_FORMAT_W(6) " messages dropped...", *counter);
+    output->write_blocking(decorations, out_of_band);
+    *counter = 0;
+  }
+
+  return true;
+}
+
+void LogAsyncFlusher::writeback(const LinkedList<AsyncLogMessage>& logs) {
+  LinkedListIterator<AsyncLogMessage> it(logs.head());
+  while (!it.is_empty()) {
+    AsyncLogMessage* e = it.next();
+    e->writeback();
+    e->destroy();
+  }
+}
+
+void LogAsyncFlusher::flush(bool with_lock) {
+  LinkedListImpl<AsyncLogMessage, ResourceObj::C_HEAP, mtLogging> logs;
+
+  if (with_lock) { // critical area
+    // Caveat: current thread must not hold _tty_lock or other lower rank lockers.
+    // Cannot install ttyUnlocker here because flush() may be invoked before defaultStream
+    // initialization.
+    MutexLocker ml(&_lock, Mutex::_no_safepoint_check_flag);
+    _buffer.pop_all(&logs);
+    AsyncLogMapIterator iter;
+    _stats.iterate(&iter);
+  } else {
+    // C++ lambda can simplify the code snippet.
+    _buffer.pop_all(&logs);
+    AsyncLogMapIterator iter;
+    _stats.iterate(&iter);
+  }
+
+  writeback(logs);
+}
+
+void LogAsyncFlusher::run() {
+  while (_state == ThreadState::Running) {
+    {
+      MonitorLocker m(&_lock, Mutex::_no_safepoint_check_flag);
+      m.wait(500 /* ms, timeout*/);
+    }
+    flush();
+  }
+
+  // Signal thread has terminated
+  MonitorLocker ml(Terminator_lock);
+  _state = ThreadState::Terminated;
+  ml.notify_all();
+}
+
+LogAsyncFlusher* LogAsyncFlusher::_instance = nullptr;
+
+void LogAsyncFlusher::initialize() {
+  if (!LogConfiguration::is_async_mode()) return;
+
+  if (_instance == NULL) {
+    Atomic::release_store(&LogAsyncFlusher::_instance, new LogAsyncFlusher());
+  }
+}
+
+// Termination
+// 1. issue an atomic release_store to close the logging window.
+// 2. flush itself in-place.
+// 3. signal asynclog thread to exit.
+// 4. wait until asynclog thread exits.
+// 5. (optional) delete this in post_run().
+void LogAsyncFlusher::terminate() {
+  if (_instance != NULL) {
+    LogAsyncFlusher* self = _instance;
+
+    Atomic::release_store<LogAsyncFlusher*, LogAsyncFlusher*>(&_instance, nullptr);
```

Comment:
> IIUC the mechanism you signal "async"-ness to the logging callsites with is via this pointer being NULL or not, right?

yes. 
> I am not sure this is safe wrt to lost log output on termination. 

It's not safe! now it may lose a coupe of messages in race condition. I have a patch to fix that. 

> Thinking this through, one could even scratch the LogAsyncBuffer dynamic allocation and make it either a static 
global object or even an AllStatic, with all static methods. 

I tried this as well. I found it's hard. Logging sites spread the entire lifecycle of JVM. Unfortunately, async logging can't be in function without some runtime supports, such as Thread::current(), threading.  One lesson I learned from this patch is that DON'T use any runtime in such low-level module.  

Now my target is to define a clear interval. Between 2 synchronization points, async logging takes over synchronous logging. I think it's acceptable because it covers the entire "java" runtime which is concerned by Java applications. 
```
|JVM begin---LogAsyncFlusher::initialize()---...---LogAsyncFlusher::terminate()--JVM End|
```

Another bonus for this 2-point-design is that we can assume only AsyncLog Thread is doing logging I/O when async logging is in charge.  it means I can skip the FileLocker imposed by unified logging framework.

Reply:
> > IIUC the mechanism you signal "async"-ness to the logging callsites with is via this pointer being NULL or not, right?
> 
> yes.
> 
> > I am not sure this is safe wrt to lost log output on termination.
> 
> It's not safe! now it may lose a coupe of messages in race condition. I have a patch to fix that.
> 
> > Thinking this through, one could even scratch the LogAsyncBuffer dynamic allocation and make it either a static
> > global object or even an AllStatic, with all static methods.
> 
> I tried this as well. I found it's hard. Logging sites spread the entire lifecycle of JVM. Unfortunately, async logging can't be in function without some runtime supports, such as Thread::current(), threading. One lesson I learned from this patch is that DON'T use any runtime in such low-level module.
> 
> Now my target is to define a clear interval. Between 2 synchronization points, async logging takes over synchronous logging. I think it's acceptable because it covers the entire "java" runtime which is concerned by Java applications.
> 
> ```
> |JVM begin---LogAsyncFlusher::initialize()---...---LogAsyncFlusher::terminate()--JVM End|
> ```

That's fine.

> 
> Another bonus for this 2-point-design is that we can assume only AsyncLog Thread is doing logging I/O when async logging is in charge. it means I can skip the FileLocker imposed by unified logging framework.

Which is good, but you have to find a solution for log sites which do not fulfill conditions to be logged asynchronously, if they fall into the live phase, see my comment above.